src/coreclr/jit/lclvars.cpp

```diff
@@ -2017,6 +2017,14 @@ bool Compiler::StructPromotionHelper::CanPromoteStructVar(unsigned lclNum)
         return false;
     }
 
+#if defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+    if (varDsc->lvIsSplit)
+    {
+        JITDUMP("  struct promotion of V%02u is disabled because it is splitted\n", lclNum);
+        return false;
+    }
+#endif // TARGET_LOONGARCH64 || TARGET_RISCV64
```

Comment:
ARM32 has split parameters and does not need this so I do not think this is the right fix.

One thing ARM32 does is to always use dependent promotion for parameters. Can you please check if making `lvaGetPromotionType` return `PROMOTION_TYPE_DEPENDENT` for split parameters is enough to fix the problem?

If not then we should understand why and how ARM32 differs in its handling so that we can align the behavior of all 3 platforms.

Reply:
The LoongArch64 is OK, this issue doesn't meet on LA64.

This is strange.

Reply:
> One thing ARM32 does is to always use dependent promotion for parameters. Can you please check if making lvaGetPromotionType return PROMOTION_TYPE_DEPENDENT for split parameters is enough to fix the problem?

I checked however it cannot fix the problem.

> The LoongArch64 is OK for the above two cases, this issue doesn't meet on LA64.

Thank you so much for sharing the test results on LA64 CHECKED build. I think there are some issues in RISC-V only. Something is missed or implemented in wrong way in RISC-V only.

I will investigate the split parameters of ARM32 and RISC-V. If you have any comments about the problem, please let me know. It will be very helpful to me.
Thank you.

Reply:
We tested these cases some times.
The console format is ok, but the cases  tested by new xunit will failed random on runtime8.0 while runtime6.0 is ok.

As the Dragon Boat Festival for Chinese, I will debug it on sunday.

Reply:
I found a difference between ARM32 and RISCV64.
ARM32 doesn't support `FEATURE_MULTIREG_STRUCT_PROMOTE`. So in ARM32, `ShouldPromoteStructVar` returns false with `Not promoting promotable struct local V%02u, because lvIsParam is true and #fields = 2` message.

https://github.com/dotnet/runtime/blob/2e764d6eb092fc8f05c42e6f137347c3eab45fd0/src/coreclr/jit/lclvars.cpp#L2158-L2194

This is my test codes with RISCV64 and ARM32.
Build as CHECKED and executed with `export COMPlus_TieredCompilation=0`
```
namespace MyApp
{
    internal ref struct Test
    {
        [MethodImplAttribute(MethodImplOptions.NoInlining)]
        private Test(int a, int b, int c, int d, int e, int f, Span<int> g, int h) // FOR RISCV64
        {
            Console.WriteLine(a);
            Console.WriteLine(b);
            Console.WriteLine(c);
            Console.WriteLine(d);                                                                                                                        
            Console.WriteLine(e);
            Console.WriteLine(f);
            Console.WriteLine(g.IsEmpty);
            Console.WriteLine(h);
        }
        [MethodImplAttribute(MethodImplOptions.NoInlining)]
        private Test(int a, int b, Span<int> c, int d) // FOR ARM32
        {
            Console.WriteLine(a);
            Console.WriteLine(b);
            Console.WriteLine(c.IsEmpty);
            Console.WriteLine(d);
        }
        [MethodImplAttribute(MethodImplOptions.NoInlining)]
        public static void check()
        {
            var a = new Test(1, 2, 3, 4, 5, 6, stackalloc int[2], 9);
            var b = new Test(1, 2, stackalloc int[2], 3);
        }
    }
    internal class Program
    {
       static void Main(string[] args)
        {
            Test.check();
            return;
        }
    }
}
```

Reply:
I see. Then the current fix seems fine to me -- but maybe move it into the `FEATURE_MULTIREG_STRUCT_PROMOTE` section you have highlighted?

Reply:
Updated. Thank you.